utils/build_script_generator.py

```diff
@@ -2,168 +2,83 @@
-                script=script_content_indent
-            )
-
-            scripts[step_id] = step_script
-
-    environment_vars = [f'export {key}="{var}"' for key, var in environment.items()]
-
-    compiled_script = template.format(
-        environment='\n'.join(environment_vars),
-        build_steps='\n\n'.join(scripts.values()),
-        workspace=current_path,
-        call_build_steps='\n'.join([f'    build_{uuid}' for uuid in scripts.keys()]),
-    )
-
-    return compiled_script
 
 def main() -> int:
-    parsers = {
-        'macos': parse_macos_job,
-        'ubuntu': (lambda x: x),
-        'windows': (lambda x: x),
-    }
-    arguments = sys.argv
-
     if len(sys.argv) > 1:
         file_name = sys.argv[1]
     else:
         file_name = ".github/workflows/main.yml"
+    try:
+        with open(file_name) as workflow_file:
+            yaml_data = yaml.load(workflow_file, Loader=yaml.FullLoader)
 
-    with open(file_name) as workflow_file:
-        yaml_data = yaml.load(workflow_file, Loader=yaml.FullLoader)
-
-        global_env = yaml_data.get('env', {})
-        jobs = yaml_data.get('jobs', {}).values()
-        job_items = [item for item in jobs if not item.get('needs', None)]
-
-        for target_os in ['macos-latest', 'windows-latest', 'ubuntu-latest']:
-            target_jobs = [item for item in job_items if target_os in item.get('runs-on', [])]
-            os_name = target_os.split('-')[0]
-
-            template_filename = f'./templates/build-script-{os_name}.tpl'
-
-            with open(template_filename, 'r') as template_file:
-                template = template_file.read()
-
-            step_template = BLUEPRINT[os_name]
+            global_env = yaml_data.get("env", {})
+            jobs = yaml_data.get("jobs", {})
 
-            for step_number, target_job in enumerate(target_jobs, start=1):
-                compiled_script = parsers[os_name](
-                    target_job, template, step_template,
-                    global_env
+            for target_os in ["macos"]:
+                build_script = f"./CI/build-deps-{target_os}.sh"
+                environment_data = jobs.get(f"{target_os}-deps-build", {}).get(
+                    "env", {}
                 )
-                file_name = f'build-script-{os_name}-{step_number:02d}.sh'
 
-                print(f' + Writing compiled script to file {file_name}')
-                with open(file_name, 'w') as script_file:
-                    script_file.write(compiled_script)
+                filtered_data = {
+                    key: value
+                    for key, value in environment_data.items()
+                    if key
+                    not in [
+                        "MACOSX_DEPLOYMENT_TARGET",
+                        "FFMPEG_REVISION",
+                        "BLOCKED_FORMULAS",
+                    ]
+                }
+
+                # Use preserved item order in dicts with Py 3.7
+                dependencies = dict.fromkeys(
```

Comment:
@derrod lmk if you know of an easier pattern to get get pairs of `A_VERSION` and `A_HASH` into a single list containing only `A` while also preserving order.. 😬

Reply:
Wouldn't a good old for-loop be simpler?

Something like
```python
dependency_strings = []

for key in filtered_data.keys():
    prefix = key.partition('_')[0]
    dependency_strings.append(
        f'    "{prefix.lower()}'
        f' {environment_data.get(f"{prefix}_VERSION", "")}'
        f' {environment_data.get(f"{prefix}_HASH", "")}"'
    )
```

A normal python list preserves order anyway so I'm not entirely sure why you're essentially creating a list from the `dict_keys` view, only to then turn that into a dict again, which then gets used in a `dict_keys` view again. The list from `[key.split("_")[0] for key in filtered_data.keys()]` could be used directly.

Reply:
Or wait, the issue is that things aren't unique, right?

Reply:
The double conversion eliminates duplicates, that's the trick. By creating a dict from the keys minus the suffix, I eliminate the duplicates, but still preserve order. `set(dict.keys()` shuffles the order, so that was out of the question.. 😂

Reply:
In that case the only thing I really see is just using a generator instead of creating a temporary list, and combining the two into one big mess. A for-loop over a dictionary will also iterate over keys by default so you don't need to explicitly do `.keys()`

```python
dependency_strings = [
    (
        f'    "{key.lower()}'
        f" {environment_data.get(f'{key}_VERSION', '')}"
        f" {environment_data.get(f'{key}_HASH', '')}\""
    )
    for key in dict.fromkeys(key.partition('_')[0] for key in filtered_data)
]
```